### PR TITLE
Replace TestUnit files with Minitest.

### PR DESCRIPTION
**Pull request by kaspth**
Rails 4 switched the test framework from the venerable TestUnit to Minitest.

However, some bits were still left over. These are addressed here.

Changes:
- Renames the test framework to minitest. You'll see "invoke minitest" when
  running a test generator.
- Renames the test framework railtie to rails/minitest/railtie.
  Deprecates rails/test_unit/railtie.
- Renames the directory holding testing bits to minitest. Also renames the
  Rails::TestUnitReporter to Rails::MinitestReporter.
- Replaces mentions of TestUnit in guides with Minitest.

### CODE REVIEW

**Comment by kaspth** [comment]
@arthurnn you proposed this in Campfire a while ago. How does it look? :grin:

**Comment by maclover7** [comment]
Small reminder to make sure we add this to the 5.0 upgrading guide. On a quick first glance, this looks really good, and should hopefully reduce confusion. :-)

File: railties/lib/rails/generators/rails/integration_test/USAGE [hunk 1]
```diff
@@ -3,7 +3,7 @@ Description:
     CamelCased or under_scored, as an argument.
 
     This generator invokes the current integration tool, which defaults to
-    TestUnit.
+    Minitest.
```
Comment (nateberkopec) [inline railties/lib/rails/generators/rails/integration_test/USAGE]: Should say "minitest", no caps.

**Comment by rafaelfranca** [comment]
hmmm. I'm not sure if we should change it. Minitest is an implementation detail of our test framework, so we should not advertise it.

**Comment by nateberkopec** [comment]
Anywhere minitest is mentioned in the documentation, it should probably read "minitest" rather than "Minitest". See their [README](https://github.com/seattlerb/minitest).

@rafaelfranca I get that, but if anyone goes looking to extend their tests or to learn more about how to use the framework and they end up in the old test/unit docs, they're going to have a Bad Time.

**Comment by rafaelfranca** [comment]
Rails test framework is different of minitest, see what we do with `refute_*` methods, also we define our own autorun, so even if people en up in the minitest documentation they may have a Bad Time. I still think we should not advertise minitest as our test framework because it is not.

**Comment by arthurnn** [comment]
> Minitest is an implementation detail of our test framework, so we should not advertise it.

Not sure if I agree. Minitest is a hard dependency, and we extend it, but in the end of the day, a vanila Rails app will have Minitest. 
Also, As far as I remember, Rspec, changes the generators to be called rspec instead of test unit.

If we dont wanna advertise Minitest, we should at least changed to something like `test` label. But I still think it is OK to embrace that hard dependency we have and call it so.

**Comment by rafaelfranca** [comment]
arel is a hard dependency too but it is an implementation detail. We don't advertise arel as public API and that people should extend it. `test_unit` is from the style of test not from the gem called `test_unit`. Minitest also support spec type tests which BTW, we disable in Rails, exactly because we want people to use the test unit approach.

File: guides/source/configuring.md [hunk 2]
```diff
@@ -172,7 +172,7 @@ Rails allows you to alter what generators are used with the `config.generators`
 ```ruby
 config.generators do |g|
   g.orm :active_record
-  g.test_framework :test_unit
+  g.test_framework :minitest
```
Comment (arthurnn) [inline guides/source/configuring.md]: is test_framework option here a implementation thing @rafaelfranca ?
Maybe we should change a few places but not all, thoughts?

**Comment by arthurnn** [comment]
https://www.google.ca/webhp?sourceid=chrome-instant&ion=1&espv=2&ie=UTF-8#q=test+unit

File: guides/source/configuring.md (hunk 2 above)
Comment (rafaelfranca) [inline guides/source/configuring.md]: This one yes. We can in the future change from `minitest` to `awesome_test` but we should not require users to change their config. In the end `test_unit` here means, the test framework included with Rails.

**Comment by nateberkopec** [comment]
:+1: If you think Minitest is an implementation detail, changing the name (in places where it makes sense) to a new name entirely is a good idea.

xUnit is, unfortunately, also the name of a few implementations of the pattern. Maybe RailsUnit? But the name itself doesn't really matter, so as long as it isn't the same as an existing framework name.

**Comment by rafaelfranca** [comment]
The name is "Rails' TestUnit". It was not TestUnit when we used TestUnit as implementation neither is minitest now that we use minitest as implementation, so like I already said, test_unit being the name of the framework inside Rails make sense.

**Comment by arthurnn** [comment]
@rafaelfranca I totally understand your point, and definitely you have more context around that than I do. I was just concerned about people been confused between a framework name and a name around the  context of Rails. To be frank I already saw people confused, specially when migrating from rails 3.2 to 4 , as that when the underline framework had changed. However, also, nobody ever complained about it, so you got a fair point. I am ok in not changing the name.

**Comment by kaspth** [comment]
Two more nails in the coffin:

  * We switched to Minitest in Rails 4 released in June 2013. That coupled with Rails 5 requiring Ruby 2.2 it's been a long time since users have had contact with TestUnit. Any confusion is likely to have been battle tested enough by now.
  * The test framework to require is `rails/test_unit/railtie`, which does make it clearer that we aren't necessarily dealing with the Ruby TestUnit framework of old.

Gonna pool resources on something else. Thanks everyone! ❤️

**Pull request closed without merging**